Don't attempt to set invalid fan percentage and preset_mode

Flameeyes opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

When loading state from a saved scene, the `percentage` and `preset_mode`,
may be set to `None`, but calling the service to set them to `None` would result
in an error (#62792.)

Instead handle these two attributes differently:

 - when turning the entity 'on' from 'off' state, call the full form of `turn_on`
   service;
 - if the fan is already on, only call the service for the attribute
   that is not `None`.

This also introduces tests to make sure that the transitions are
behaving as expected.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #62792
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jbouwh commented:
Integrations should ignore a `None` state for percentage or preset_mode when this is requested through a service call. Only the integration itself can reset the attribute state. Is this the intended behavior for this PR?

Flameeyes commented:
> Integrations should ignore a `None` state for percentage or preset_mode when this is requested through a service call. Only the integration itself can reset the attribute state. Is this the intended behavior for this PR?

That sounds a lot like the question from @frenck in https://github.com/home-assistant/core/pull/53484/files#r676563499

The answer is that the schema of the service call does not allow `None` to be passed to the service call.

Flameeyes commented:
Any better ideas about this one?

allenporter commented on tests/components/fan/test_reproduce_state.py:
```diff
@@ -85,3 +94,181 @@ async def test_reproducing_states(hass, caplog):
+
+
+async def test_modern_turn_on_invalid(hass, caplog):
+    """Test correct reproduction of fan states when using preset modes."""
```
Double check these test comments as a few of them appear to be copy/paste

allenporter commented on homeassistant/components/fan/reproduce_state.py:
```diff
@@ -33,14 +33,23 @@
     ATTR_PRESET_MODE: SERVICE_SET_PRESET_MODE,
 }
 
+ATTRIBUTES = set(SIMPLE_ATTRIBUTES.keys()) | set(SPEED_AND_MODES_ATTRIBUTES)
```
My impression is that `set` may not preserve order, which i worry about given the comment above:
```
>>> a = {'a': 2, 'b': 3, 'c': 4}
>>> a.keys()
dict_keys(['a', 'b', 'c'])
>>> set(a.keys())
{'b', 'a', 'c'}
```

Is that a problem?

allenporter commented on homeassistant/components/fan/reproduce_state.py:
```diff
@@ -73,18 +82,39 @@ async def _async_reproduce_state(
-                data[attr] = state.attributes[attr]
-                service_calls[service] = data
-
+            # Turn on the fan with all the speed and modes settings.
```
Can you help me follow/understand why `SIMPLE_ATTRIBUTES` is not used when going from off to on?

allenporter commented on homeassistant/components/fan/reproduce_state.py:
```diff
@@ -73,18 +82,39 @@ async def _async_reproduce_state(
+                    service_calls[service] = data
+
+            # Speed and modes are only changed if they are non-None.
+            # That's because they are mutually exclusive, but we always
```
I'm having trouble following what we're trying to say here. Can you perhaps think of a way to rephrase? I get what you mean by them being mutually exclusive, but i don't understand what that has to do with being non-none or what "we always store all of them" means and why that matters.

Flameeyes replied:
Let me know if the way I wrote it now makes sense.

allenporter commented on homeassistant/components/fan/reproduce_state.py:
```diff
@@ -73,18 +82,39 @@ async def _async_reproduce_state(
+                    data[attr] = state.attributes[attr]
+            service_calls[SERVICE_TURN_ON] = data
+        else:
+            # The simple attributes are copied directly.
```
Why is it ok for these to have `None` values but not the others? Naively, i just read the method for `def set_direction(self, direction: str) -> None:` and it doesn't appear to accept a non-value.

allenporter replied:
If the fan does not support the feature, can the service call be omitted? Or do we require calling the set call with a None value in some cases?

Flameeyes replied:
Technically, yes we can omit the service call if the feature is not supported. The reason why calling the services even when not supported is okay is because in their schema definition, the values are marked as optional: https://github.com/home-assistant/core/blob/dev/homeassistant/components/fan/__init__.py#L126

Your question had me wonder if I can rely on just the fact that, when changing one, the other _should_ turn to None (unless the integration is broken), but then I realized that this does not help us here because the `cur_state` is a snapshot at the beginning of the restore, not updated live, so we can't rely on a sequence such as off → on(percentage) → `state.preset_mode == cur_state.preset_mode == None`.

If instead of batching the service calls we would request to act immediately, that would be possible… but I don't think that's desired.

allenporter reviewed (commented):
I'm not super familiar with fan, so i have some naive questions below. This is also subtle, so just looking to make sure i deeply understand what is happening here and why the specifics of the approach were taken.

(vs naively, not sure why we can't just add a null check in the existing loop, for example)

Flameeyes commented:
The test failure on 3.10 seems to be completely unrelated to my change 🤷‍♂️

allenporter commented on homeassistant/components/fan/reproduce_state.py:
```diff
@@ -32,14 +32,23 @@
     ATTR_PRESET_MODE: SERVICE_SET_PRESET_MODE,
 }
 
+ATTRIBUTES = set(SIMPLE_ATTRIBUTES.keys()) | set(SPEED_AND_MODES_ATTRIBUTES)
```
I think this can be removed if we remove the  `Return if we are already at the right state.` check at the top. It seems like the code below should be a no-op if all the attributes are the same so i'm not sure its worth preserving given the extra code being added, and lets us remove this variable.

Flameeyes replied:
There was one small change needed to maintain behaviour. I've fixed that now.

allenporter commented on homeassistant/components/fan/reproduce_state.py:
```diff
@@ -72,18 +81,46 @@ async def _async_reproduce_state(
+            # turn the fan on.
+            data = service_data.copy()
+            for attr in SPEED_AND_MODES_ATTRIBUTES:
+                if attr in state.attributes and state.attributes[attr] is not None:
```
Given the None check is repeated a few times, I think it makes sense to just do this first at the top of the section before updating attributes, e.g.:
```
state_attributes = {k: v for k, v in state.attributes.items() if v is not None}
```

allenporter replied:
Is the `check_attr_equal` check OK to run here too? or is it avoided intentionally? I think it may not be needed in this case, but it seems like if it doesn't hurt then we can combine all attribute filters together.

allenporter commented on homeassistant/components/fan/reproduce_state.py:
```diff
@@ -72,18 +81,46 @@ async def _async_reproduce_state(
+                if (
+                    attr in state.attributes
+                    and state.attributes[attr] is not None
+                    and not check_attr_equal(
```
Similar to the suggestion for the `None` check above, lets pull out a variable for the attributes that differences, e.g., something like this:
```
updated_attributes = {k: v for k,v in state_attributes.items() if state_attributes.get(k) != v }
```

Then that can be used to simplify building the updates.

allenporter replied:
(I'm not sure we even need `check_attr_equal` given its a pretty trivial check)

allenporter commented on homeassistant/components/fan/reproduce_state.py:
```diff
@@ -72,18 +81,46 @@ async def _async_reproduce_state(
+            # them is always going to be stored as None. If we were to
+            # try to set it, it will raise an error. So instead we
+            # only set the one that is non-None.
+            for attr, service in SPEED_AND_MODES_ATTRIBUTES.items():
```
My impression is that this loop is the same as the loop below, so lets combine them and just have the conditional logic update the set of eligible services to update.

One way to do this could be to have something like this:
```
update_services = {
    **(SIMPLE_ATTRIBUTES if cur_state.state == STATE_ON else {}),
    **SPEED_AND_MODES_ATTRIBUTES,
}
for attr, service in update_services:
    ...
```
or just build up the set of services to update using the existing conditional logic already here and use the loop below.

allenporter reviewed (commented):
Thanks for all the clarifications. I think I understand the goals now, and i'd like to try to make this code a bit more compact. I think this is really close, and so now we're just optimizing, so that we can follow the logic a little easier. There are a few instances of similar looking blocks of code, so what we can do is try to factor out the similarities so we can better understand the differences in place.

Flameeyes commented:
I tried simplifying the code and reducing copy-paste in a slightly different way. Among other this this moves adding the entity_id later in the code, that shouldn't be any more expensive, but most importantly is easier to read the state change above.

I also changed `service_calls` to be a list of tuples rather than a dictionary because at first it wasn't obvious to me it was actually counting on the order of insertion — if that's not desirable I can revert that part.

bdraco commented:
https://github.com/home-assistant/core/pull/59781 removed the legacy speed compatibility shim.  The speeds themselves are due to be removed any time now.  It might be best to wait on this until after that happens.

Flameeyes commented:
This will still be needed because `percentage` and `preset_mode` are still mutually exclusive.

bdraco commented:
https://github.com/home-assistant/core/pull/67743

Flameeyes commented:
Updated now that @bdraco's change is in.

frenck commented:
Why is this handled in the reproduce state? And not as part of the main entity model?

Flameeyes commented:
> Why is this handled in the reproduce state? And not as part of the main entity model?

I followed what the climate state restore as implemented in #53484 — I'm not sure how this can be handled otherwise.

frenck commented:
We handle this in, e.g., the select entity (setting an option that doesn't exist). I think that makes more sense, as it protects its on a higher level (and thus also benefits e.g., service calls).

Flameeyes commented:
I'm not sure I follow.

The `select` service raises if the option selected is not valid: https://github.com/home-assistant/core/blob/dev/homeassistant/components/select/__init__.py#L53 and the reproduce state explicitly guards against calling the service with an invalid option: https://github.com/home-assistant/core/blob/dev/homeassistant/components/select/reproduce_state.py#L29

In this case, we're not even dealing with an invalid state. There's two possible, legal state for the fan to be represented as: `{'percentage': value, 'preset_mode': None}` or `{'percentage': None, 'preset_mode': value}`.

I'm probably missing something, but I see two options to solve this:

 1. the current pull request: we explicitly use the full `turn_on` service when going off→on, so that the entity model and specific component deal with it, and we only set the one of the two attributes that is not None if the state is on→on;
 2. only add to the state the one of the two attributes that is not None, so that only the call to set the right one is sent, and no special casing is needed on the attributes.

The reason why I think 1 is better is because a) it doesn't require changing the scenes and b) it does not go into off→on(previous speed)→on(saved speed) steps, since it uses the fully specified `turn_on`.

allenporter commented on homeassistant/components/fan/reproduce_state.py:
```diff
@@ -56,37 +61,51 @@ async def _async_reproduce_state(
+    for service, data in service_calls:
         await hass.services.async_call(
-            DOMAIN, service, data, context=context, blocking=True
+            DOMAIN, service, dict(data, **service_data), context=context, blocking=True
```
Optional suggestion: It seems like this might be slightly easier to read if this is inlined like:
```
{
    ATTR_ENTITY_ID: state.entity_id,
    **data
}
```

The motivation for the suggestion is there are just a few variables here to track (`service`, `data`, and `service_data`). However, what is here is perfectly correct, so consider this optional as it may be a style preference.

frenck commented:
> The select service raises if the option selected is not valid: [https://github.com/home-assistant/core/blob/dev/homeassistant/components/select/__init__.py#L53](https://github.com/home-assistant/core/blob/dev/homeassistant/components/select/__init__.py?rgh-link-date=2022-03-11T14%3A21%3A24Z#L53) and the reproduce state explicitly guards against calling the service with an invalid option: [https://github.com/home-assistant/core/blob/dev/homeassistant/components/select/reproduce_state.py#L29](https://github.com/home-assistant/core/blob/dev/homeassistant/components/select/reproduce_state.py?rgh-link-date=2022-03-11T14%3A21%3A24Z#L29)

Yeah not fully agreeing on that one either. It has partially the same problem caused here,
applying an invalid scene by just ignoring what is invalid. That gives the idea everything is good and working and all, while it is not.

IMHO, if a scene (or any service call for that matter), contains invalid data, stuff should fail. Not silently ignore and continue.

Flameeyes commented:
> IMHO, if a scene (or any service call for that matter), contains invalid data, stuff should fail. Not silently ignore and continue.

I agree! But in this case we're having valid scenes that fail: any scene containing a fan that supports preset modes currently fail because one of the two attributes is saved in the state as None.

Indeed, my first proposed solution was #59688 which did exactly that option 2: only stored the values that would have to be called in the fan itself. If we go that route, users with (valid, but currently non-functioning) scenes including preset mode capable fans will have to update their scenes, which is not the worst thing but it is a bit annoying — but to be honest, right now the only way for those scenes to work is to manually edit the scene file to do exactly that, so it'll still be an improvement.

I think there's still some logic needed for this in restore state though, to avoid the fan turning on with the previous setting and _then_ changing to whatever it was meant to. I can do that as well, if that's preferred.

allenporter commented:
Are we saying the `None` value should not be set in the state that needs to be reproduced in the first place? (e.g. based on current supported modes.)

I'm trying to remember if a similar solution would have worked in
https://github.com/home-assistant/core/pull/53484#discussion_r676563499 but maybe there was a difference due to how properties aren't just based on supported features, but intertwined with hvac modes -- which perhaps is not the case here.

Flameeyes commented:
> Are we saying the `None` value should not be set in the state that needs to be reproduced in the first place? (e.g. based on current supported modes.

I would have gone on whether the attribute is None, rather than based on supported features, because…

> I'm trying to remember if a similar solution would have worked in [#53484 (comment)](https://github.com/home-assistant/core/pull/53484#discussion_r676563499) but maybe there was a difference due to how properties aren't just based on supported features, but intertwined with hvac modes -- which perhaps is not the case here.

… fan is not as bad as climate, but it has a similar problem: `direction` and `oscillate` are simply optional based on supported features — `percentage` and `preset_mode` are a bit more complicated:

 * as long as `SUPPORT_SET_SPEED` is present, `percentage` attribute is supported, but it may be None if `preset_mode` is not None;
 * as long as both `SUPPORT_SET_SPEED` and `SUPPORT_PRESET_MODE` are both present, `preset_mode` attribute is supported, but it may be None if `percentage` is not None.

Which means that if we gate the presence of attributes in the state _solely_ on the supported features, we would end up with the following valid situations:

 * neither `percentage` nor `preset_mode`;
 * `percentage` (non-None), but no `preset_mode`;
 * `percentage != None and preset_mode == None`
 * `percentage == None and preset_mode != None`

allenporter commented:
> > Are we saying the `None` value should not be set in the state that needs to be reproduced in the first place? (e.g. based on current supported modes.
> 
> I would have gone on whether the attribute is None, rather than based on supported features, because…
> 
> > I'm trying to remember if a similar solution would have worked in [#53484 (comment)](https://github.com/home-assistant/core/pull/53484#discussion_r676563499) but maybe there was a difference due to how properties aren't just based on supported features, but intertwined with hvac modes -- which perhaps is not the case here.
> 
> … fan is not as bad as climate, but it has a similar problem: `direction` and `oscillate` are simply optional based on supported features — `percentage` and `preset_mode` are a bit more complicated:
> 
> * as long as `SUPPORT_SET_SPEED` is present, `percentage` attribute is supported, but it may be None if `preset_mode` is not None;
> * as long as both `SUPPORT_SET_SPEED` and `SUPPORT_PRESET_MODE` are both present, `preset_mode` attribute is supported, but it may be None if `percentage` is not None.
> 
> Which means that if we gate the presence of attributes in the state _solely_ on the supported features, we would end up with the following valid situations:
> 
> * neither `percentage` nor `preset_mode`;
> * `percentage` (non-None), but no `preset_mode`;
> * `percentage != None and preset_mode == None`
> * `percentage == None and preset_mode != None`

Was the idea of the discussion here about not setting it in the scene rather than setting it with `None`?

Flameeyes committed:
fan: when restoring state from off, use the turn_on parameters.

This simplifies the state restore function by using the full parameters
of the fan.turn_on service, which allows skipping a transition, if the
"off" speed of the fan would be different than desired.

This also includes a full set of tests for the various state transitions
to make sure that fans can be turned on, off, or change parameters no
matter what the current state is.

Flameeyes committed:
fan: only store one of percentage or preset_mode attributes.

Since the two attributes are mutually exclusive, when a fan entity
supports both preset modes and percentage speeds, one of the two
attributes has to be None (setting preset_mode should set percentage to
None and vice-versa), and cannot be reproduced.

This allows simplifying the state reproduction state code by assuming
all of the stored attributes are safe to reproduce.

Flameeyes commented:
Updated to latest dev.

@allenporter @frenck does it make sense as it is?

allenporter commented on homeassistant/components/fan/reproduce_state.py:
```diff
@@ -56,37 +62,45 @@ async def _async_reproduce_state(
+                (
+                    SERVICE_TURN_ON,
+                    {
+                        attr: state.attributes.get(attr)
```
While this may iterate over more attributes, perhaps it is more natural as it is more directly showing that we're taking a subset of the existing state attributes and applying them:

```
{ k: v for k, v in state.attributes if k in SPEED_AND_MODE_ATTRIBUTES }
```

How about renaming `SPEED_AND_MODE_ATTRIBUTES` to `TURN_ON_ATTRIBUTES`?

allenporter replied:
```
                    {
                        attr: state.attributes[attr]
                        for attr in TURN_ON_ATTRIBUTES
                        if attr in state.attributes
                    }
```

allenporter commented on homeassistant/components/fan/reproduce_state.py:
```diff
@@ -56,37 +62,45 @@ async def _async_reproduce_state(
-
-    service_data = {ATTR_ENTITY_ID: state.entity_id}
-    service_calls = {}  # service: service_data
+    service_calls: list[tuple[str, dict[str, Any]]] = []
```
I think the `service_calls` format can stay how it is as a dict, since dicts preserve insertion order by default.  This will just make the diff a bit smaller

allenporter commented on homeassistant/components/fan/reproduce_state.py:
```diff
@@ -56,37 +62,45 @@ async def _async_reproduce_state(
+        # Attributes that are not supported will not show up in the state.
+        # Percentage and preset mode attributes would already be set, but
+        # this simplifies the code in exchange for a no-op call.
         for attr, service in ATTRIBUTES.items():
```
Will this loop over the speed and mode attributes again?

allenporter commented:
> Updated to latest dev.
> 
> @allenporter @frenck does it make sense as it is?

This makes sense to me.  There are really two separate changes here now:
- Fix state so that it is persisted properly
- Streamline the turn on behavior

How about we split these into two independent changes, each smaller, since they are not really related anymore, and should make it easier to quickly get them each in.

allenporter commented:
> > Updated to latest dev.
> > @allenporter @frenck does it make sense as it is?
> 
> This makes sense to me. There are really two separate changes here now:
> 
> * Fix state so that it is persisted properly
> * Streamline the turn on behavior
> 
> How about we split these into two independent changes, each smaller, since they are not really related anymore, and should make it easier to quickly get them each in.

(I'll review these each quickly, apologies for the delay)

frenck commented on homeassistant/components/fan/__init__.py:
```diff
@@ -347,13 +347,14 @@ def state_attributes(self) -> dict:
 
         if supported_features & FanEntityFeature.SET_SPEED:
-            data[ATTR_PERCENTAGE] = self.percentage
+            if self.percentage is not None:
+                data[ATTR_PERCENTAGE] = self.percentage
```
This will cause `percentage` to be missing when its currently unknown. This is not correct, we have to make sure the key is present at any time.

For context see: <https://github.com/home-assistant/architecture/discussions/680#discussion-3724991>

allenporter replied:
Does this mean that it's a valid state for the attributes to be None and thus it's correct to handle on the side where state is reproduced? (The initial direction of this PR) Or is there another layer of indirection somewhere else when translating this to a saved scene?

Flameeyes replied:
Yeah I'm stuck at this point.

https://github.com/home-assistant/core/pull/62794#issuecomment-1065914641 tells me to not ignore "invalid Scenes" where `percentage: None; preset_mode: str` or `percentage: int; preset_mode: None`.

https://github.com/home-assistant/core/pull/62794#discussion_r846772928 tells me to always have the key present at any time (sometimes as `None`).

I'm more than happy to keep working on the style issues, and figure out how to implement this cleanly, but the fact of the matter is that fan entities with preset modes cannot be added and reproduced onto scenes following the documented design of the fan entity — and I'm afraid I don't have enough time to get involved into redesigning either the Scene handling or the fan entity.

I'd be happy to put a bounty on it if anyone wants to do that! I don't care to be the one fixing fan entities with Scenes, I just want fan entities to work with Scenes!

bdraco replied:
I'm a bit late to the party but I'll try to establish some understanding so we can move this forward:


> [#62794 (comment)](https://github.com/home-assistant/core/pull/62794#issuecomment-1065914641) tells me to not ignore "invalid Scenes" where 

> `percentage: None; preset_mode: str` 

In this state, the fan itself is in control of the percentage since the preset mode ie Auto, etc is controlling it.  The percentage is solely determined by the device and we can never restore it since we don't control it.

> `percentage: int; preset_mode: None`.

In this state the fan is not controlling the percentage and the remote/Home Assistant/physical device/etc is in control.  We can restore the percentage but there is no preset mode set so nothing to restore.

> 
> [#62794 (comment)](https://github.com/home-assistant/core/pull/62794#discussion_r846772928) tells me to always have the key present at any time (sometimes as `None`).


> https://github.com/home-assistant/core/pull/62794#discussion_r846772928 tells me to always have the key present at any time (sometimes as None).

The key should be there if its supported as integrations rely on it.

Flameeyes replied:
> > `percentage: None; preset_mode: str`
> 
> In this state, the fan itself is in control of the percentage since the preset mode ie Auto, etc is controlling it. The percentage is solely determined by the device and we can never restore it since we don't control it.

Correct.

> 
> > `percentage: int; preset_mode: None`.
> 
> In this state the fan is not controlling the percentage and the remote/Home Assistant/physical device/etc is in control. We can restore the percentage but there is no preset mode set so nothing to restore.

Correct.

> > [#62794 (comment)](https://github.com/home-assistant/core/pull/62794#discussion_r846772928) tells me to always have the key present at any time (sometimes as `None`).
> 
> > [#62794 (comment)](https://github.com/home-assistant/core/pull/62794#discussion_r846772928) tells me to always have the key present at any time (sometimes as None).
> 
> The key should be there if its supported as integrations rely on it.

But that's the thing: those two situations above *cannot ever be restored* — integrations can't rely on them, because the  `fan.set_percentage` and `fan.set_preset_mode` services will reject calls with `None` before they ever reach any integration.

So when restoring a Scene that has `percentage: None; preset_mode: str`, the list of calls currently includes `set_percentage(percentage=None); set_preset(preset_mode='Something')` — the first service call will fail, and the state restore is aborted midway.

That's why my (second) attempt at fixing this with this pull request I changed it to only restore _one_ of the two attributes. I changed due to the comments here to go back to my original suggestion of not storing the None-valued-attribute, since setting `percentage: int` or `preset_mode: str` implies the mutual-exclusive attribute to `None`, but that's what's being rejected here.

bdraco replied:
> But that's the thing: those two situations above _cannot ever be restored_ — integrations can't rely on them, because the `fan.set_percentage` and `fan.set_preset_mode` services will reject calls with `None` before they ever reach any integration.

We can't ever call the services with `None` as these are effectively unknown values.

> 
> So when restoring a Scene that has `percentage: None; preset_mode: str`, the list of calls currently includes `set_percentage(percentage=None); set_preset(preset_mode='Something')` — the first service call will fail, and the state restore is aborted midway.

We shouldn't call it with `None` since the fan is in control and we can't restore what we don't have control of.

> 
> That's why my (second) attempt at fixing this with this pull request I changed it to only restore _one_ of the two attributes. 

That makes sense.  Either we are in control or the fan is in control, its mutually exclusive.

> I changed due to the comments here to go back to my original suggestion of not storing the None-valued-attribute, since setting `percentage: int` or `preset_mode: str` implies the mutual-exclusive attribute to `None`, but that's what's being rejected here.

I think we should still store them all and only restore the one we can control

Flameeyes replied:
> I think we should still store them all and only restore the one we can control

That works for me (basically, going back to 83892741d6f2eb053d513311c844db4244d61db2.)

If that works for @allenporter and @frenck too, I'm happy to make this change tonight/tomorrow.

bdraco replied:
With all the force pushes, I can't really comment on https://github.com/home-assistant/core/commit/83892741d6f2eb053d513311c844db4244d61db2 

Do you have a branch I could checkout to review?

Flameeyes replied:
I created a separate pull request (#70294) with a combination of the change from this diff (to use `turn_on` with additional attributes) plus the original change on this request to just ignore either `percentage` or `preset_mode` if they are `None`.

bdraco replied:
I think the version in https://github.com/home-assistant/core/pull/70294 is fine.  It doesn't change the fan entity and handles the mutually exclusive case.

I marked this PR as a draft as i didn't want to close it since it has all the history of how we arrived at #70294

frenck commented on homeassistant/components/fan/__init__.py:
```diff
@@ -347,13 +347,14 @@ def state_attributes(self) -> dict:
             supported_features & FanEntityFeature.PRESET_MODE
             or supported_features & FanEntityFeature.SET_SPEED
-        ):
+        ) and self.preset_mode is not None:
```
Same as above.
